### ISSUE TITLE
apple2cp: crash to monitor observed after keypress in A2Desktop

### ISSUE THREAD

**Issue by inexorabletash** [mamedev/mame#7640]
Reported (by me) at: 

https://github.com/a2stuff/a2d/issues/358

Repro: 
* `mame64 apple2cp -skip_gameinfo -window -nomax -flop3 A2DeskTop_800k.2mg` - [disk image from here](https://github.com/a2stuff/a2d/releases/tag/v1.2-prealpha23)
* Once the desktop has appeared, press a key

System beeps, stops responding to mouse events. If you type `C051` then <kbd>return</kbd> the system will enter text mode, revealing a crash to monitor. The crash is not always instant; sometimes a "glitched" mouse pointer can be seen to move a bit prior to the crash.

No repro for me on real hardware (caveat: with ROM5X and booting off a Floppy Emu) or other Apple models.

Potentially useful data: A2Desktop is somewhat unusual (for Apple II software) in that it spends a lot of time executing out of Aux memory; the application logic lives in the main 64k bank, but the graphics and windowing library executes out of aux 64k, so the app is particularly sensitive to emulation bugs around banking.

Tested using 0.227 (via https://github.com/ksherlock/ample wrapper) on macOS Catalina.

cc: @rb6502

**Comment by rb6502** [comment]
Thanks!  I've reproduced the issue, but I have no idea what's going on yet :)

Interestingly it also happens on the plain IIc, but not on the IIgs.

**Comment by inexorabletash** [comment]
Odd... I just tried it on the IIc (`apple2c`) and don't repro it. Using 140k floppy image with PRODOS added. (I've updated the package script to do that for future builds)

apple2e and apple2ee also seem fine for me (unless I put a mouse card in, but that's a different issue)

**Comment by rb6502** [comment]
Ahh.  It happens on the last ROM IIc (apple2c4) but not the original.  That's a handy clue.

Is the IIe mouse card an A2D problem or a MAME one?  I know older A2D versions worked with it in MAME.

**Comment by inexorabletash** [comment]
Re: Mouse - seems to be a MAME issue; slot scan finds the mouse, locks up on the INITMOUSE call into the firmware. Works fine on real hardware, Virtual II and AppleWin. `mame64 apple2ee -skip_gameinfo -window -nomax -sl2 mouse -flop1 /A2DeskTop_140k_disk1.po -nothrottle` File a separate issue here? Or do you prefer compat problems with an app to be in a single bug regardless of root cause?

**Comment by rb6502** [comment]
That's definitely a regression, I tested A2D 1.1 and Dazzle Draw with the mouse card not too long ago (maybe a year ago or so) and it was fine.  Now it locks up waiting on some handshake from the 68HC05, it appears.  Making it a separate issue here is good.

**Comment by rb6502** [comment]
So for the A2D thing, the crash is because an RTS at $40B0 in auxram is going to $D015 in the aux language card, without the secondary $Dxxx page banked in.  Normally there's a stub there that switches back to the main bank (STA RDMAINRAM, STA WRMAINRAM, RTS).  When the crash occurs, the secondary $Dxxx page *is* banked in and everything goes wrong because there's no valid code there.

Still working out why that happens.  Might have to pull in qkumba, but I'll attempt further debugging myself first :)

**Comment by peterferrie** [comment]
> Still working out why that happens.  Might have to pull in qkumba, but I'll
> attempt further debugging myself first :)

Sure.  Let me know. :-)

**Comment by rb6502** [comment]
Fixed in https://github.com/mamedev/mame/commit/403f7a099af070e5a5eca1c7742d01933c94a774

Code in the firmware was checking RdRAM2 at $C011 to determine how to restore the LC state and if a key was down, well, you can see bit 7 was being set because it was acting like the $C010 return value ORed with LCBank2.  That only happened on the IIc and IIc Plus, and apparently older versions of the firmware didn't care.
